Quit gracefully if Bitcoin RPC connection lost

AdamISZ opened the pull request:
Fixes #442.
First, the CONNREFUSED socket error is handled in jsonrpc.
Second, we respond to this (but *not* to resets) with a reactor
shutdown in BitcoinCoreInterface.rpc(). This also necessitates
early-quitting in the calling function
(WalletService.transaction_monitor) since the reactor stop
will only stop future deferred calls, not the currently running
one. The obvious sys.exit approach is only used in startup,
because the reactor is not currently running at that point.
Also minor change to DummyBlockchainInterface for test.

undeath commented:
imo the desired behaviour would be to try reconnecting to bitcoind instead of quitting.

AdamISZ commented:
A little extra colour:

First, this is what it looks like when bitcoind is killed after the yield generator has started running:

```
2020-05-24 16:58:27,402 [INFO]  joined: #joinmarket-pit tor.darkscience.net
2020-05-24 16:58:27,402 [DEBUG]  topicUpdated:  tor.darkscience.net #joinmarket-pit JoinMarket open outcry pit | Discussion in #joinmarket | 0 5 LATEST RELEASE v0.6.2 Update ASAP! https://github.com/JoinMarket-Org/joinmarket-clientserver/releases/
2020-05-24 16:58:27,402 [INFO]  All IRC servers connected, starting execution.
2020-05-24 16:58:27,403 [INFO]  JM daemon setup complete
2020-05-24 17:00:00,677 [ERROR]  Failure of RPC connection to Bitcoin Core. Application cannot continue, shutting down.
2020-05-24 17:00:00,679 [INFO]  Lost IRC connection to: guybrush.hackint.org . Should reconnect automatically soon.
2020-05-24 17:00:00,681 [INFO]  Lost IRC connection to: tor.darkscience.net . Should reconnect automatically soon.
```
(the last two messages are obviously sub-ideal but I'm OK with that).

The message if you *start up* with a connection refusal is similar (that will be more common, result of a misconfigured RPC setup).

```
User data location: /home/user/.joinmarket/
2020-05-24 16:56:11,636 [DEBUG]  rpc: getblockchaininfo []
2020-05-24 16:56:11,637 [ERROR]  Connection refused.
2020-05-24 16:56:11,637 [ERROR]  Failure of RPC connection to Bitcoin Core. Application cannot continue, shutting down.
```

AdamISZ commented:
> imo the desired behaviour would be to try reconnecting to bitcoind instead of quitting.

Well clearly that's preferable, but two points: this is *only* for connection *refused* errors, broken pipes, resets and even wrong protocol, we retry. But if the connection is refused, there is a significant possibility that we have a misconfiguration. What I'm not sure about is: what would have to be done (nothing?) in the rest of the codebase to ensure that whatever loops are still running, are not negatively affected by the fact that the blockchain is updating? I made a note about it [here](https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/585/files#diff-79627d21bc41aff521fe45af1ee3968aR193-R194)

AdamISZ commented:
Hmm, now you've forced me to think about it, it sounds less difficult than I first imagined. We could just have an infinite loop in the rpc call that keeps going as long as there's a CONNREFUSED. still have to bother with the problem of SystemExit getting caught in twisted (BareException) but that problem is kind of already addressed anyway.

undeath commented:
> But if the connection is refused, there is a significant possibility that we have a misconfiguration.

At startup that's correct. But it's a litte annoying having to restart JM every time bitcoind is restarted for whatever reason.

Maybe not an infinite loop, but a couple of minutes (something like 5-15min) until JM finally gives up would be nice. That would probably mean that RPC calls would either have to block or return a non-termminal error during that unconnected time.

AdamISZ commented:
> That would probably mean that RPC calls would either have to block or return a non-termminal error during that unconnected time.

Yes, what I had in mind was a while loop within BitcoinCoreInterface.rpc which would effectively block the entire reactor. I *guess* that should work OK, albeit there may be edge cases where something goes wrong. About 5-15 minutes instead of indefinitely, I'm not really sure.

AdamISZ commented:
Hmm I guess that would block the GUI, not OK (in itself).

undeath commented:
Just blocking reactor seems like a bad idea. Wouldn't it also cause IRC connections to time out?

AdamISZ commented:
> Wouldn't it also cause IRC connections to time out?

Yes that had also occurred to me but it seemed the existing connection recovery code would handle that.
But the GUI blocking is not OK.
The reason completely blocking the reactor made sense to me was as per the comment I linked. I'm worried about having to reason through what would happen elsewhere in the code if the wallet is not being updated and rpc calls are not working. It seems like a big mess. So I was attracted to the fact that completely blocking the reactor, by definition, freezes all operations.

kristapsk commented:
Tested, works as expected with cli apps.

Does not terminate joinmarket-qt if you shutdown node while it's running, ends up with endless loop of messages about shutting down in terminal. 
![image](https://user-images.githubusercontent.com/4500994/82830747-3ac2ab80-9ebf-11ea-803e-45ed31001eec.png)
OTOH, it seems to autorecover when you restart Bitcoin node (tested with deposit to jm wallet address after restarting node). :) With exception of following message in terminal when you are closing qt app.
```
Traceback (most recent call last):
  File "joinmarket-qt.py", line 1375, in closeEvent
    self.reactor.stop()
  File "/home/user/git/joinmarket-clientserver/jmvenv/lib/python3.6/site-packages/twisted/internet/base.py", line 647, in stop
    "Can't stop reactor that isn't running.")
twisted.internet.error.ReactorNotRunning: Can't stop reactor that isn't running.
```

Also, both if there's start-up failure in joinmarket-qt or it happens during runtime, all errors are outputted to terminal, which also aren't ideal (Qt message box would be better).

kristapsk commented:
> But it's a litte annoying having to restart JM every time bitcoind is restarted for whatever reason.

This at least can be monitored by a simple shell script that checks is yield generator process is running and sends you an e-mail or message through some other channel if it is not, so you can manually restart almost immediately. Definitely an improvement.

AdamISZ commented:
Yes the whole Qt aspect was left unconsidered. I'm still not actually sure the right road with all this. I will do my best to come back to it as quickly as possible, thanks to both for your thoughts and testing. If you have any other ideas about it feel free to jot them down here.

kristapsk commented:
GUI can be made to terminate immediately the same way as cli with this patch:
```diff
diff --git a/jmclient/jmclient/blockchaininterface.py b/jmclient/jmclient/blockchaininterface.py
index 0d6657e..cb9d406 100644
--- a/jmclient/jmclient/blockchaininterface.py
+++ b/jmclient/jmclient/blockchaininterface.py
@@ -1,5 +1,6 @@
 
 import abc
+import os
 import random
 import sys
 import time
@@ -211,9 +212,7 @@ class BitcoinCoreInterface(BlockchainInterface):
             # BareException type).
             log.error("Failure of RPC connection to Bitcoin Core. "
                       "Application cannot continue, shutting down.")
-            if reactor.running:
-                reactor.stop()
-            return None
+            os._exit(EXIT_FAILURE)
         return res
 
     def is_address_labeled(self, utxo, walletname):
```
But I don't think terminating GUI without GUI message is ideal, it will look like a random crash for some users. Especially as I do believe we all agree that having one click GUI in the future would be a good thing to have. Also, GUI seems to be recovering from this (although I haven't tested what happens if node's RPC stops responding in middle of single join or tumbler run).

Also, before this PR, with initial config error or Bitcoin node not running at start-up you got some, although kinda cryptic, error message in the GUI.
![image](https://user-images.githubusercontent.com/4500994/82842844-c51d0680-9ee3-11ea-9019-0426fcdf0eb2.png)

AdamISZ commented:
As I said above:

> Yes the whole Qt aspect was left unconsidered

This isn't complete for a couple of reasons, and Qt behaviour is one. The other one is the uncertainties discussed above.

AdamISZ commented:
Rebased on master as of https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/585/commits/5ce4b55d53cb5f22ebed6a422a7c79f6326d5a03

Added a commit to handle the Qt quit gracefully with https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/585/commits/cc219ccba10ef493ef9a8a3e31ad676ec8f7fbef

This seems to work fine from my tests, just a dialog box saying that the wallet service fails and it's probably due to Core RPC failure, then application quits.

Obviously allowing the recovery when the connection returns is more desirable, but I don't have a good solution to that. And to emphasize once again, this will *only* be happening where the connection is literally refused/shut down, not broken pipe etc.

kristapsk commented:
It still does not give any GUI feedback if you start JoinMarketQt with RPC connection not working, error is printed only in terminal. If one starts JoinMarketQt with an icon click not from terminal he will not see any feedback why app is not starting. It is worse than before these changes.

Also, it does nothing in GUI if you terminate node with no wallet loaded. Then when trying to load wallet with node not running, you get some very cryptic error message for ordinary user. And clicking "Ok" does not terminate app. But if you then restart your node and try loading wallet again it does recover and works :).
![image](https://user-images.githubusercontent.com/4500994/84422413-a5951600-ac25-11ea-9268-0f285f32507c.png)

If full node is terminated with wallet loaded it behaves as expected.

AdamISZ commented:
@kristapsk thanks for the careful testing.

Well this turned out to be *extremely* fiddly and annoying, overall, even to achieve the most basic possible goal. Please read the commit message of https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/585/commits/8840ebfbd3126ebb76b8444e1272601159a59bad as I think it explains in detail the final changes I've made.

I now see the behaviour as at least acceptable in all cases: the 3 Qt cases mentioned and the yieldgenerator case. Yes, it could be better, but this has already taken a pretty large chunk of time.

kristapsk commented on jmclient/jmclient/wallet_service.py:
```diff
@@ -40,15 +40,28 @@ def __init__(self, wallet):
+                # that bci is now failing when we are starting
+                # the wallet service.
+                raise Exception("WalletService failed to start "
+                                "due to inability to query block height.")
```
Append "is your node running?" here too?

kristapsk commented on jmclient/jmclient/wallet_service.py:
```diff
@@ -73,14 +86,36 @@ def update_blockheight(self):
+        has been correctly updated).
         """
+
+        def critical_error():
```
Rename this to `get_blockheight_error()`?

AdamISZ replied:
I mean it's scoped only within the function, so I'd feel it's fine even to use `error()`. I think I'll leave it, just to avoid another commit.

kristapsk commented:
Apart from review comments, ACK, re-tested all the things I had mentioned in comments above about GUI, also re-tested that yield generator quits when node is stopped.

AdamISZ commented:
Thanks again @kristapsk . I'm sorry the end result isn't perfect, but it's just very difficult, for whatever reason.
